pkg/ingress/mcp/generator_test.go

```diff
@@ -0,0 +1,146 @@
+package mcp
+
+import (
+	"testing"
+
+	"github.com/golang/protobuf/ptypes"
+	extensions "istio.io/api/extensions/v1alpha1"
+	mcp "istio.io/api/mcp/v1alpha1"
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config"
+)
+
+func TestGenerate(t *testing.T) {
+	tests := []struct {
+		name      string
+		fn        func() *model.PushContext
+		generator model.McpResourceGenerator
+		isErr     bool
+	}{
+		{
+			name: "VirtualService",
+			fn: func() *model.PushContext {
+				ctx := model.NewPushContext()
+				cfg := config.Config{
+					Spec: &networking.VirtualService{},
+				}
+				ctx.AllVirtualServices = []config.Config{cfg}
+				return ctx
+			},
+			generator: VirtualServiceGenerator{},
+			isErr:     false,
+		},
+		{
+			name: "Gateway",
+			fn: func() *model.PushContext {
+				ctx := model.NewPushContext()
+				cfg := config.Config{
+					Spec: &networking.Gateway{},
+				}
+				ctx.AllGateways = []config.Config{cfg}
+				return ctx
+			},
+			generator: GatewayGenerator{},
+			isErr:     false,
+		},
+		{
+			name: "EnvoyFilter",
+			fn: func() *model.PushContext {
+				ctx := model.NewPushContext()
+				cfg := config.Config{
+					Spec: &networking.EnvoyFilter{},
+				}
+				ctx.AllEnvoyFilters = []config.Config{cfg}
+				return ctx
+			},
+			generator: EnvoyFilterGenerator{},
+			isErr:     false,
+		},
+		{
+			name: "DestinationRule",
+			fn: func() *model.PushContext {
+				ctx := model.NewPushContext()
+				cfg := config.Config{
+					Spec: &networking.DestinationRule{},
+				}
+				ctx.AllDestinationRules = []config.Config{cfg}
+				return ctx
+			},
+			generator: DestinationRuleGenerator{},
+			isErr:     false,
+		},
+		{
+			name: "WasmPlugin",
+			fn: func() *model.PushContext {
+				ctx := model.NewPushContext()
+				cfg := config.Config{
+					Spec: &extensions.WasmPlugin{},
+				}
+				ctx.AllWasmplugins = []config.Config{cfg}
+				return ctx
+			},
+			generator: WasmpluginGenerator{},
+			isErr:     false,
+		},
+		{
+			name: "WasmPlugin with wrong config",
+			fn: func() *model.PushContext {
+				ctx := model.NewPushContext()
+				cfg := config.Config{
+					Spec: "string",
+				}
+				ctx.AllWasmplugins = []config.Config{cfg}
+				return ctx
+			},
+			generator: WasmpluginGenerator{},
+			isErr:     true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, _, err := test.generator.Generate(nil, test.fn(), nil, nil)
+			if (err != nil && !test.isErr) || (err == nil && test.isErr) {
+				t.Errorf("Failed to generate config: %v", err)
+			}
+		})
+	}
+}
+
+func TestMarshal(t *testing.T) {
+	ctx := model.NewPushContext()
+	cfg := config.Config{
+		Spec: &networking.VirtualService{
```

Comment:
A meaningful test is to unmarshal the mcp.body obtained by marshal back to virtualservice and compare it with the original object